prow/spyglass/gcsartifact_test.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// Tests reading at most n bytes of data from files in GCS
+func TestGCSReadAtMost(t *testing.T) {
+	buildLogArtifact := NewGCSArtifact(context.Background(), fakeGCSBucket.Object("logs/example-ci-run/403/build-log.txt"), "", "build-log.txt", 500e6)
+	testCases := []struct {
+		name     string
+		artifact viewers.Artifact
+		n        int64
+		expected []byte
+	}{
+		{
+			name:     "ReadN example build log",
+			n:        4,
+			artifact: buildLogArtifact,
+			expected: []byte("Oh "),
+		},
+	}
+	for _, tc := range testCases {
+		actualBytes, err := tc.artifact.ReadAtMost(tc.n)
+		if err != nil {
+			t.Errorf("Test %s failed with err:\n%s", tc.name, err)
+			continue
+		}
+		if !bytes.Equal(actualBytes, tc.expected) {
+			t.Errorf("Test %s failed.\nExpected: %s\nActual: %s", tc.name, tc.expected, actualBytes)
+		}
+	}
+}
+
+// Tests reading all data from files in GCS
+func TestGCSReadAll(t *testing.T) {
```

Comment:
Right now these tests are at the highest level, where you need:
 - connectivity to GCS (or your fake)
 - no issues reading data
 - correct logic to read all / read at most / read N/ whatever

Maybe it would help to refactor a tiny bit and allow you to unit-test that your logic works even with a fake that is much less complicated than the GCS fake you are using. Instead of using a fake GCS server, use dependency injection to add a closure that returns a specific `[]byte` on the call you're expecting, and error otherwise, etc. Reduce the scope of moving parts of these tests. The top-level test as you have it is useful, still, but harder to reason about

Reply:
added a working dependency injection for `*storage.ObjectHandle` in latest